netty/src/main/java/io/grpc/transport/netty/WriteQueue.java

```diff
@@ -0,0 +1,162 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import com.google.common.base.Preconditions;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelPromise;
+
+import java.util.ArrayDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A queue of pending writes to a {@link Channel} that is flushed as a single unit.
+ */
+class WriteQueue {
+
+  private static final int DEQUE_CHUNK_SIZE = 128;
+
+  /**
+   * {@link Runnable} used to schedule work onto the tail of the event loop.
+   */
+  private final Runnable later = new Runnable() {
+    @Override
+    public void run() {
+      flush();
+    }
+  };
+
+  private final Channel channel;
+  private final BlockingQueue<Runnable> queue;
+  private final AtomicBoolean scheduled = new AtomicBoolean();
+
+  /**
+   * ArrayDeque to copy queue into when flushing in event loop.
+   */
+  private final ArrayDeque<Runnable> writeChunk =
+      new ArrayDeque<Runnable>(DEQUE_CHUNK_SIZE);
+
+  public WriteQueue(Channel channel) {
+    this.channel = Preconditions.checkNotNull(channel, "channel");
+    queue = new LinkedBlockingQueue<Runnable>();
+  }
+
+  /**
+   * Schedule a flush on the channel.
+   */
+  void scheduleFlush() {
+    if (scheduled.compareAndSet(false, true)) {
+      // Add the queue to the tail of the event loop so writes will be executed immediately
+      // inside the event loop. Note DO NOT do channel.write outside the event loop as
+      // it will not wake up immediately without a flush.
+      channel.eventLoop().execute(later);
+    }
+  }
+
+  /**
+   * Enqueue a write command on the channel.
+   *
+   * @param command a write to be executed on the channel.
+   * @param flush true if a flush of the write should be schedule, false if a later call to
+   *              enqueue will schedule the flush.
+   */
+  void enqueue(Object command, boolean flush) {
+    enqueue(command, channel.newPromise(), flush);
```

Comment:
I tried using VoidPromise as an experiment earlier and it didn't work. Rather than working through the issues I figured it something that could be done in a follow up.